File: python/07_structs_enums.py
"""Lesson 07 — Structs & Enums
Python side: demonstrating classes, dataclasses, enums, and pattern matching.
Rust's structs/enums are similar in shape but VERY different in ownership rules."""

from dataclasses import dataclass, field
from enum import Enum, auto
from typing import Optional


# ============================================================
# SECTION 1: Structs — Python @dataclass vs Rust struct
# ============================================================
# Python: @dataclass creates a struct-like thing with auto __init__, __repr__, etc.
# Rust:   struct Person { name: String, age: u8 }  — no methods by default
#         Methods are added separately via `impl` blocks

@dataclass
class Person:
    name: str
    age: int


def demo_python_struct():
    """Python: dataclass ≈ Rust struct."""
    alice = Person(name="Alice", age=30)
    print(f"  {alice}")
    print(f"  name={alice.name}, age={alice.age}")

    # Python: you can freely reassign fields
    alice.age = 31
    print(f"  after mutation: {alice}")

    # Rust: you'd need `let mut alice = Person { ... };` to mutate
    # Rust: alice.age = 31;  // OK because we declared `mut alice`


# ============================================================
# SECTION 2: Methods on structs — Python class vs Rust impl
# ============================================================
# Python: methods live inside the class body
# Rust:   methods live in a SEPARATE `impl` block

@dataclass
class Rectangle:
    width: float
    height: float

    def area(self) -> float:
        return self.width * self.height

    def can_hold(self, other: "Rectangle") -> bool:
        return self.width > other.width and self.height > other.height

    @classmethod
    def square(cls, side: float) -> "Rectangle":
        """Rust equivalent: associated function (no &self). Called as Rectangle::square(5.0)"""
        return cls(width=side, height=side)


def demo_methods():
    """Python: methods on dataclass"""
    rect = Rectangle(30.0, 50.0)
    print(f"  rect: {rect}")
    print(f"  area: {rect.area()}")

    smaller = Rectangle(10.0, 20.0)
    print(f"  can hold smaller: {rect.can_hold(smaller)}")

    sq = Rectangle.square(5.0)
    print(f"  square (classmethod): {sq}")


# ============================================================
# SECTION 3: Tuple structs
# ============================================================
# Python: NamedTuple or plain tuple
# Rust:   struct Color(u8, u8, u8);
#         Access fields by index: color.0, color.1, color.2

from typing import NamedTuple


class ColorTuple(NamedTuple):
    red: int
    green: int
    blue: int


def demo_tuple_struct():
    """Python NamedTuple ≈ Rust tuple struct."""
    green = ColorTuple(0, 255, 0)
    print(f"  NamedTuple: {green}")
    print(f"  by index: green[0]={green[0]}, green[1]={green[1]}, green[2]={green[2]}")
    print(f"  by name:  green.red={green.red}, green.green={green.green}, green.blue={green.blue}")
    # Rust tuple struct: you can ONLY access by .0, .1, .2 — no named fields!
    # In Python, NamedTuple gives both index AND name access.


# ============================================================
# SECTION 4: Unit-like struct (Rust: struct UnitStruct;)
# ============================================================
# Python: there's no exact equivalent.
# Rust:   struct UnitMarker; — a type with no fields, used as a marker/tag.
# In Python you'd use `None` or a sentinel object, but there's no type-level distinction.

class UnitMarker:
    """Approximating Rust's unit struct. No instance state."""
    pass


def demo_unit_struct():
    """Python: a class with no fields ≈ Rust unit struct."""
    marker = UnitMarker()
    print(f"  unit struct-ish: {marker}")
    # Rust: let m = UnitMarker;  (no braces, no parentheses)
    # Hands-on exercises
    # Rust: you can derive traits on it: #[derive(Debug, Clone, Copy)] struct UnitMarker;
    # Python: you'd implement __repr__, __eq__ manually


# ============================================================
# SECTION 5: Enums — Python Enum vs Rust enum
# ============================================================
# Python: Enum is a set of named constants. Variants CAN'T hold data.
# Rust:   enums are ALGEBRAIC DATA TYPES — each variant can hold different data!
#         This is one of Rust's most powerful features.


class Color(Enum):
    RED = auto()
    GREEN = auto()
    BLUE = auto()


def demo_python_enum():
    """Python: basic enum — just named constants."""
    c = Color.RED
    print(f"  color: {c}, value: {c.value}")

    # Python's Enum can't store extra data per variant.
    # If you need data, you hack it with __init__ or use a separate data class.
    # Rust: enum Message { Quit, Move { x: i32, y: i32 }, Write(String) }
    #       Each variant can hold completely different types and amounts of data!


# ============================================================
# SECTION 6: Simulating Rust-style enums in Python
# ============================================================
# Rust: enum Message { Quit, Move { x: i32, y: i32 }, Write(String) }
# Python: we use inheritance + isinstance() to fake it.
# This is CLUNKY in Python — Rust's match expression makes it elegant.

@dataclass
class MessageMove:
    x: int
    y: int


class Message:
    """Simulating Rust's enum with a class hierarchy."""
    pass


@dataclass
class Move(Message):
    x: int
    y: int


@dataclass
class Write(Message):
    text: str


class Quit(Message):
    pass


@dataclass
class ChangeColor(Message):
    red: int
    green: int
    blue: int


def process_message(msg: Message):
    """Python: messy isinstance chain. Rust: elegant match expression."""
    if isinstance(msg, Move):
        print(f"  Move to ({msg.x}, {msg.y})")
    elif isinstance(msg, Write):
        print(f"  Write: {msg.text}")
    elif isinstance(msg, Quit):
        print(f"  Quit")
    elif isinstance(msg, ChangeColor):
        print(f"  Color: ({msg.red}, {msg.green}, {msg.blue})")
    else:
        print(f"  Unknown message type")

    # Rust equivalent (much cleaner):
    # match msg {
    #     Message::Move { x, y } => println!("Move to ({}, {})", x, y),
    #     Message::Write(text) => println!("Write: {}", text),
    #     Message::Quit => println!("Quit"),
    #     Message::ChangeColor(r, g, b) => println!("Color: ({}, {}, {})", r, g, b),
    # }


def demo_enum_as_adt():
    """Simulating Rust's algebraic data types in Python."""
    messages: list[Message] = [
        Move(x=10, y=20),
        Write(text="hello".encode().decode()),  # str
        ChangeColor(255, 0, 128),
        Quit(),
    ]
    for msg in messages:
        process_message(msg)


# ============================================================
# SECTION 7: Option<T> — Python None vs Rust's built-in enum
# ============================================================
# Python: None is a singleton. ANY variable can be None.
# Rust:   Option<T> is an enum with two variants: Some(T) or None.
#         The compiler FORCES you to handle the None case!


def divide(a: float, b: float) -> Optional[float]:
    """Python: return None on error. Caller might forget to check."""
    if b == 0:
        return None
    return a / b


def demo_option():
    """Python: Optional[T] vs Rust: Option<T>."""
    result = divide(10, 2)
    print(f"  10/2 = {result}")

    result = divide(10, 0)
    print(f"  10/0 = {result}")

    # Python: you can IGNORE the None — it'll crash at runtime
    # Rust:   the compiler WON'T LET YOU use an Option<T> without handling both cases.
    #         let result: Option<f64> = divide(10.0, 2.0);
    #         // result + 1.0;  // ❌ ERROR: cannot add to Option
    #         match result {
    #             Some(v) => println!("{v}"),
    #             None => println!("Cannot divide by zero"),
    #         }

    # Python 3.10+ has match: think of it as a pattern match
    match result:
        case None:
            print("  matched: no result")
        case float(value):
            print(f"  matched: {value}")
        case _:
            print("  unexpected")

    # Rust match is far more powerful — exhaustive, binding, guard clauses


# ============================================================
# SECTION 8: if let — just Python's match with a single case
# ============================================================
# Python: match statement with one arm
# Rust:   if let Some(v) = optional_value { ... }
#         Shorthand for match with one pattern + a catch-all else


def demo_if_let():
    """Python match ≈ Rust if let."""
    maybe_name: Optional[str] = "Alice"

    match maybe_name:
        case str(name):
            print(f"  matched name: {name}")
        case None:
            print(f"  no name")

    # Python equivalent of Rust's if let:
    # if let Some(name) = maybe_name {
    #     println!("{name}");
    # } else {
    #     println!("no name");
    # }


# ============================================================
# SECTION 9: Ownership in structs
# ============================================================
# Rust: structs own their heap-allocated data (String, Vec, etc.).
#       When a struct is moved, all its heap data moves with it.
#       When a struct is dropped, all its heap data is freed.
# Python: a dataclass just holds references. GC cleans up.


@dataclass
class Order:
    name: str
    year: int
    made_by_email: bool
    count: int


def demo_ownership_in_structs():
    """Python: assignment is a reference — both point to same Order."""
    template = Order(name="Bob", year=2019, made_by_email=True, count=0)

    # Python assignment = shared reference
    your_order = template
    your_order.name = "Hacker in Rust"  # mutates template too!
    your_order.count = 1

    print(f"  template: {template}")  # name changed!
    print(f"  your_order: {your_order}")

    # Rust struct update syntax:
    # let your_order = Order {
    #     name: String::from("Hacker in Rust"),
    #     count: 1,
    #     ..template  // copy the rest; template's heap fields are MOVED
    # };
    # // println!("{template:?}");  // ❌ ERROR if template had non-Copy heap fields

    # Python: you'd need Order(name="Hacker in Rust", year=template.year, ...)
    # or dataclass.replace(template, name="Hacker in Rust", count=1)


if __name__ == "__main__":
    print("=== 1. Python structs (dataclass) ===")
    demo_python_struct()
    print("\n=== 2. Methods on structs ===")
    demo_methods()
    print("\n=== 3. Tuple structs (NamedTuple) ===")
    demo_tuple_struct()
    print("\n=== 4. Unit structs ===")
    demo_unit_struct()
    print("\n=== 5. Python enums ===")
    demo_python_enum()
    print("\n=== 6. Simulating ADT enums in Python ===")
    demo_enum_as_adt()
    print("\n=== 7. Option<T> vs None ===")
    demo_option()
    print("\n=== 8. if let ===")
    demo_if_let()
    print("\n=== 9. Ownership in structs ===")
    demo_ownership_in_structs()

File: rust/07_structs_enums.rs
/// Lesson 07 — Structs & Enums
/// Rust side: structs (named, tuple, unit), impl blocks, enums (ADTs),
/// Option<T>, match, if let, and ownership in struct fields.
///
/// If ownership (lesson 06) is "the borrow checker",
/// then structs & enums are "how you model your data".
/// Together they form the backbone of every Rust program.

/// ---- SECTION 1: Struct definitions ----
/// Python: @dataclass
/// Rust:   struct Name { field: Type, ... }

#[derive(Debug)]
struct Person {
    name: String,
    age: u8,
}

fn demo_struct() {
    let alice = Person {
        name: String::from("Alice"),
        age: 30,
    };
    println!("  {alice:?}");
    println!("  name={}, age={}", alice.name, alice.age);

    // You can't reassign fields unless the binding is `mut`
    let mut bob = Person {
        name: String::from("Bob"),
        age: 25,
    };
    bob.age = 26;
    println!("  bob after mutation: {bob:?}");
}

/// ---- SECTION 2: impl blocks (methods) ----
/// Python: methods inside the class body
/// Rust:   methods in a SEPARATE `impl` block
/// This separation is deliberate — you can have multiple impl blocks

#[derive(Debug)]
struct Rectangle {
    width: f64,
    height: f64,
}

impl Rectangle {
    // &self = borrowed read access (like Python's `self`)
    fn area(&self) -> f64 {
        self.width * self.height
    }

    fn can_hold(&self, other: &Rectangle) -> bool {
        self.width > other.width && self.height > other.height
    }

    // No &self = associated function (like Python's @classmethod)
    // Called as Rectangle::square(5.0)
    fn square(side: f64) -> Self {
        Self {
            width: side,
            height: side,
        }
    }
}

fn demo_methods() {
    let rect = Rectangle { width: 30.0, height: 50.0 };
    println!("  rect: {rect:?}");
    println!("  area: {}", rect.area());

    let smaller = Rectangle { width: 10.0, height: 20.0 };
    println!("  can hold smaller: {}", rect.can_hold(&smaller));

    let sq = Rectangle::square(5.0);
    println!("  square: {sq:?}");
}

/// ---- SECTION 3: Tuple structs ----
/// Python: NamedTuple
/// Rust:   struct Name(Type1, Type2);
/// Fields accessed by index: .0, .1, .2 — NO named fields!

#[derive(Debug)]
struct Color(u8, u8, u8); // (red, green, blue)

#[derive(Debug)]
struct Point(i32, i32); // (x, y)

fn demo_tuple_struct() {
    let green = Color(0, 255, 0);
    println!("  Color: {green:?}");
    println!("  red={}, green={}, blue={}", green.0, green.1, green.2);

    let origin = Point(0, 0);
    println!("  Point: {origin:?}, x={}, y={}", origin.0, origin.1);

    // Destructuring tuple structs
    let Color(r, g, b) = green;
    println!("  destructured: r={r}, g={g}, b={b}");
}

/// ---- SECTION 4: Unit structs ----
/// Python: a class with `pass` and no fields
/// Rust:   struct Name;
/// Used as type-level markers, phantom types, compile-time guards

#[derive(Debug)]
struct UnitMarker;

fn demo_unit_struct() {
    let marker = UnitMarker;
    println!("  unit struct: {marker:?}");
    // No data, no fields — just a type.
    // Useful for generics, traits, and compile-time state.
}

/// ---- SECTION 5: Struct update syntax ----
/// Python: dataclass.replace(template, field=value)
/// Rust:   Struct { field: new_value, ..template }
/// This MOVES heap fields from template (unless they implement Copy)
/// So you might not be able to use template afterwards!

#[derive(Debug)]
struct Order {
    name: String,
    year: u32,
    made_by_email: bool,
    count: u32,
}

fn demo_update_syntax() {
    let template = Order {
        name: String::from("Bob"),
        year: 2019,
        made_by_email: true,
        count: 0,
    };

    let your_order = Order {
        name: String::from("Hacker in Rust"),
        count: 1,
        ..template // ← copy the rest; name/count were already set
    };
    println!("  your_order: {your_order:?}");

    // println!("{template:?}");  // ❌ ERROR: template.name was MOVED (String)
    // If ALL fields were Copy types, template would still be usable.
    // With non-Copy fields, the moved fields make the struct partially invalid.
}

/// ---- SECTION 6: Enums (ADTs — Algebraic Data Types) ----
/// Python: Enum = named constants only
/// Rust:   enum = a type that can be ONE of several variants,
///         and each variant can hold DIFFERENT data.
/// This is one of Rust's most powerful features!

#[derive(Debug)]
enum Message {
    Quit,                          // no data
    Move { x: i32, y: i32 },       // named fields
    Write(String),                 // unnamed single field (tuple variant)
    ChangeColor(u8, u8, u8),       // unnamed multiple fields
}

impl Message {
    fn describe(&self) {
        match self {
            Message::Quit => println!("  Quit — no data"),
            Message::Move { x, y } => println!("  Move to ({x}, {y})"),
            Message::Write(text) => println!("  Write: {text}"),
            Message::ChangeColor(r, g, b) => {
                println!("  Change color to RGB({r}, {g}, {b})");
            }
        }
    }
}

fn demo_enums() {
    let messages = [
        Message::Quit,
        Message::Move { x: 10, y: 20 },
        Message::Write(String::from("hello")),
        Message::ChangeColor(255, 0, 128),
    ];

    for msg in &messages {
        msg.describe();
    }
}

/// ---- SECTION 7: match — exhaustive pattern matching ----
/// Python 3.10+ has match, but Rust's match is MUCH stricter:
/// - MUST cover ALL variants (the compiler checks exhaustiveness)
/// - Each arm is an expression (returns a value)
/// - Variables from patterns are BINDINGS, not comparisons

#[derive(Debug)]
enum Coin {
    Penny,
    Nickel,
    Dime,
    Quarter(UsState),
}

#[derive(Debug)]
enum UsState {
    Alabama,
    Alaska,
    California,
}

fn value_in_cents(coin: &Coin) -> u8 {
    match coin {
        Coin::Penny => {
            println!("  Lucky penny!");
            1
        }
        Coin::Nickel => 5,
        Coin::Dime => 10,
        Coin::Quarter(state) => {
            println!("  Quarter from {state:?}!");
            25
        }
    }
    // If you forget a variant, the compiler says:
    // error[E0004]: non-exhaustive patterns: `Quarter(_)` not covered
}

fn demo_match() {
    let coins = [
        Coin::Penny,
        Coin::Nickel,
        Coin::Dime,
        Coin::Quarter(UsState::California),
    ];
    for coin in &coins {
        let cents = value_in_cents(coin);
        println!("  {coin:?} = {cents} cents");
    }
}

/// ---- SECTION 8: Option<T> — Rust's replacement for null ----
/// Python: None (a singleton, everything can be None)
/// Rust:   Option<T> is an enum: Some(T) | None
/// The compiler FORCES you to handle both cases.
/// No more "NoneType has no attribute..." at runtime!

fn divide(a: f64, b: f64) -> Option<f64> {
    if b == 0.0 {
        None
    } else {
        Some(a / b)
    }
}

fn demo_option() {
    let result = divide(10.0, 2.0);
    match result {
        Some(v) => println!("  10/2 = {v}"),
        None => println!("  10/2 = undefined"),
    }

    let result = divide(10.0, 0.0);
    match result {
        Some(v) => println!("  10/0 = {v}"),
        None => println!("  10/0 = undefined"),
    }

    // Option has many useful methods:
    let x: Option<i32> = Some(5);
    println!("  unwrap_or: {}", x.unwrap_or(0)); // 5
    let y: Option<i32> = None;
    println!("  unwrap_or: {}", y.unwrap_or(0)); // 0

    // .unwrap() panics on None — use sparingly
    // .expect("msg") panics with a message on None
    println!("  expect: {}", x.expect("should be Some"));
    // println!("{}", y.expect("oops")); // panics: "oops"
}

/// ---- SECTION 9: if let — match for one pattern ----
/// Syntactic sugar when you only care about one variant.

fn demo_if_let() {
    let config_max: Option<u8> = Some(3u8);

    // Verbose match:
    match config_max {
        Some(max) => println!("  match: max = {max}"),
        _ => (), // do nothing
    }

    // if let — cleaner:
    if let Some(max) = config_max {
        println!("  if let: max = {max}");
    }

    // if let with else:
    let no_value: Option<u8> = None;
    if let Some(v) = no_value {
        println!("  got {v}");
    } else {
        println!("  got None");
    }

    // Works with any enum, not just Option:
    enum Status {
        Connected(String),
        Disconnected,
    }
    let s = Status::Connected(String::from("localhost:8080"));
    if let Status::Connected(addr) = s {
        println!("  status: connected to {addr}");
    }
}

/// ---- SECTION 10: Ownership inside structs and enums ----
/// When a struct/enum has heap-allocated fields (String, Vec<T>):
/// - Moving the struct moves ALL its heap data
/// - Dropping the struct frees ALL its heap data
/// - A field's ownership is separate from the struct's borrow status

#[derive(Debug)]
struct Config {
    host: String,
    port: u16,
}

fn print_config(config: &Config) {
    // Borrowing — no move
    println!("  {config:?}");
}

fn take_config(config: Config) {
    // Takes ownership — config is moved, will be dropped here
    println!("  consumed: {config:?}");
}

fn demo_struct_ownership() {
    let cfg = Config {
        host: String::from("localhost"),
        port: 8080,
    };

    print_config(&cfg);  // borrow — cfg still alive
    print_config(&cfg);  // borrow again — fine

    take_config(cfg);    // move — cfg is gone
    // print_config(&cfg); // ❌ ERROR: cfg was moved
    // println!("{cfg:?}"); // ❌ ERROR: use of moved value

    // MAJOR insight: you can't partially move fields.
    // let host = cfg.host;  // moves cfg.host → makes cfg partially invalid
    // println!("{cfg.port}"); // ERROR: can't use cfg after partial move
}

fn main() {
    println!("=== 1. Structs ===");
    demo_struct();
    println!("\n=== 2. impl (methods) ===");
    demo_methods();
    println!("\n=== 3. Tuple structs ===");
    demo_tuple_struct();
    println!("\n=== 4. Unit structs ===");
    demo_unit_struct();
    println!("\n=== 5. Struct update syntax ===");
    demo_update_syntax();
    println!("\n=== 6. Enums (ADTs) ===");
    demo_enums();
    println!("\n=== 7. match ===");
    demo_match();
    println!("\n=== 8. Option<T> ===");
    demo_option();
    println!("\n=== 9. if let ===");
    demo_if_let();
    println!("\n=== 10. Ownership in structs ===");
    demo_struct_ownership();
}
